app/lib/vacuum/statuses_vacuum.rb

```diff
@@ -41,7 +41,9 @@ def analyze_statuses!
   end
 
   def remove_from_account_conversations(statuses)
-    Status.where(id: statuses.select(&:direct_visibility?).map(&:id)).includes(:account, mentions: :account).each(&:unlink_from_conversations)
+    Status.where(id: statuses.select(&:direct_visibility?).map(&:id)).includes(:account, mentions: :account).each do |status|
+      status.send(:unlink_from_conversations)
```

Comment:
`analyze_statuses!` is never used. It should actually be removed.

https://github.com/casaper/mastodon/blob/main/app/lib/vacuum/statuses_vacuum.rb#L39-L41

Reply:
> Has calling the private method with `.each(&:unlink_from_conversations)` actually worked? (I mean with the original code?)

No, and that's precisely what this PR fixes.

> But why do we need to call this private method anyway?

Because we are using `delete_all` instead of `destroy_all`. I assume that is for performance reasons, although the whole method is weird and seems to instantiate `Status` objects anyway.

Reply:
> I assume that is for performance reasons, although the whole method is weird and seems to instantiate Status objects anyway.

I was guessing that too, but that failed on multiple layers:

1. statuses_scope causes the need of reinstantiating the batches again completely with `.select(:id, :visibility)`
2. statuses_scope does a subselect instead of `joins(:account)` 
3. remove_from_account_conversations maps the ids from the array, and does a new query on the model, with joins this time `.includes(:account, mentions: :account)`
4. remove_from_search_index maps out the ids again
5. the status model is queried again to perform delete_all
6. and delete_all is called with mapping out the ids again

This Class vacumes just as inefficient as it can get :wink:

Reply:
```rb
[1] pry(main)> Status.unscoped.kept.where(account: Account.remote).explain
  Status Load (3.5ms)  SELECT "statuses".* FROM "statuses" WHERE "statuses"."deleted_at" IS NULL AND "statuses"."account_id" IN (SELECT "accounts"."id" FROM "accounts" WHERE "accounts"."domain" IS NOT NULL)
=> EXPLAIN for: SELECT "statuses".* FROM "statuses" WHERE "statuses"."deleted_at" IS NULL AND "statuses"."account_id" IN (SELECT "accounts"."id" FROM "accounts" WHERE "accounts"."domain" IS NOT NULL)
                                           QUERY PLAN
------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.29..16.50 rows=1 width=296)
   ->  Index Scan using index_statuses_20190820 on statuses  (cost=0.14..8.16 rows=1 width=296)
   ->  Index Scan using accounts_pkey on accounts  (cost=0.14..8.16 rows=1 width=8)
         Index Cond: (id = statuses.account_id)
         Filter: (domain IS NOT NULL)
```

Reply:
Also that `Status.arel_table[:id].lt(retention_period_as_id)` is IMHO redundant:

```rb
[32] pry(main)> Status.unscoped.kept.joins(:account).merge(Account.remote).where(Status.arel_table[:id].lt(retention_period_as_id)).explain
  Status Load (1.5ms)  SELECT "statuses".* FROM "statuses" INNER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."deleted_at" IS NULL AND "accounts"."domain" IS NOT NULL AND "statuses"."id" < 109382851428352000
=> EXPLAIN for: SELECT "statuses".* FROM "statuses" INNER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."deleted_at" IS NULL AND "accounts"."domain" IS NOT NULL AND "statuses"."id" < 109382851428352000
                                           QUERY PLAN
------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.29..16.50 rows=1 width=296)
   ->  Index Scan using index_statuses_20190820 on statuses  (cost=0.14..8.16 rows=1 width=296)
         Index Cond: (id < '109382851428352000'::bigint)
   ->  Index Scan using accounts_pkey on accounts  (cost=0.14..8.16 rows=1 width=8)
         Index Cond: (id = statuses.account_id)
         Filter: (domain IS NOT NULL)
(6 rows)
```

This does the same, just as good, but explicitly, without arel_table use:

```rb
[33] pry(main)> Status.unscoped.kept.joins(:account).merge(Account.remote).where(statuses: { id: ...retention_period_as_id }).explain
  Status Load (0.9ms)  SELECT "statuses".* FROM "statuses" INNER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."deleted_at" IS NULL AND "accounts"."domain" IS NOT NULL AND "statuses"."id" < $1  [["id", 109382864338944000]]
=> EXPLAIN for: SELECT "statuses".* FROM "statuses" INNER JOIN "accounts" ON "accounts"."id" = "statuses"."account_id" WHERE "statuses"."deleted_at" IS NULL AND "accounts"."domain" IS NOT NULL AND "statuses"."id" < $1 [["id", 109382864338944000]]
                                           QUERY PLAN
------------------------------------------------------------------------------------------------
 Nested Loop  (cost=0.29..16.50 rows=1 width=296)
   ->  Index Scan using index_statuses_20190820 on statuses  (cost=0.14..8.16 rows=1 width=296)
         Index Cond: (id < '109382864338944000'::bigint)
   ->  Index Scan using accounts_pkey on accounts  (cost=0.14..8.16 rows=1 width=8)
         Index Cond: (id = statuses.account_id)
         Filter: (domain IS NOT NULL)
(6 rows)
```

Reply:
We still target Ruby 2.6 (though that may change in the short-term future), which doesn't have support for that range notation.

Reply:
1. you just need to make the range manually `nil...retention_period_as_id`
2. or if even that wouldn't work: `where('statuses.id < ?', retention_period_as_id)`

In any case,  that is readable, and not against the ORM. Safer, faster, more readable...

Reply:
BTW:

That ruby 2.6 thing seems to be broken:

```sh
$ ruby -v
ruby 2.6.9p207 (2021-11-24 revision 67954) [arm64-darwin21]

# kaspi @ mmax in ~/Projects/github/mastodon on git:main x [18:59:56]
$ bundle
/Users/kaspi/.rbenv/versions/2.6.9/lib/ruby/gems/2.6.0/gems/bundler-2.2.33/lib/bundler/vendor/thor/lib/thor/error.rb:105: warning: constant DidYouMean::SPELL_CHECKERS is deprecated
Calling `DidYouMean::SPELL_CHECKERS.merge!(error_name => spell_checker)' has been deprecated. Please call `DidYouMean.correct_error(error_name, spell_checker)' instead.
Fetching gem metadata from https://rubygems.org/.........
better_html-2.0.1 requires ruby version >= 2.7.0, which is incompatible with the current version, ruby 2.6.9p207
```

And the CI tests also don't seem to test with Ruby 2.6 any more:

https://github.com/casaper/mastodon/blob/main/.circleci/config.yml#L201-L209

Reply:
Good catch. I'll open a PR to drop 2.6 support, and this one can be revisited afterwards.

As for your other remarks, a rewrite of this class is welcome, but I think this would be a separate PR, this one is just meant to fix a crash.

Reply:
@ClearlyClaire 

> Good catch. I'll open a PR to drop 2.6 support, and this one can be revisited afterwards.

I'm not sure for how much longer Ruby 2.6 is still maintained. Probably not for very long.  
That means I agree that it is likely for the best to lift the minimum Ruby version to 2.7.

> As for your other remarks, a rewrite of this class is welcome, but I think this would be a separate PR, this one is just meant to fix a crash.

Yea, I was going out of your scope here. I didn't mean those comments as PR change requests, but I can see that I might have been not clear enough.

Reply:
Perhaps this class should never have made it into the code base in the first place, as it is now.
Obviously, there was no test coverage for that method, for example.

Reply:
@ClearlyClaire 

I created a refactor PR, addressing what I stated here

https://github.com/mastodon/mastodon/pull/21487

I only addressed the db query problems.

Because I'm unsure about unwanted side effects of the refactoring, I kept:

- `delete_all`
- `.send(:unlink_from_conversations)` – Added a TODO comment though

But the db queries should be more efficient that way, I think.
Also fixed some code readability nits, e.g. scattered code and dead code.